packages/react-reconciler/src/ReactFiberNewContext.new.js

```diff
@@ -244,12 +247,20 @@ export function propagateContextChange<T>(
           }
           scheduleWorkOnParentPath(fiber.return, renderLanes);
 
-          // Mark the updated lanes on the list, too.
-          list.lanes = mergeLanes(list.lanes, renderLanes);
-
-          // Since we already found a match, we can stop traversing the
-          // dependency list.
-          break;
+          // Mark the updated lanes on the list, too, so that the consumer
+          // knows it received an update. Unless this dependency is associated
+          // with a hook, in which case we'll let the hook decide whether to
+          // bail out when we visit it.
+          // TODO: We could call the selector right here, during propagation.
+          // That would give us the opportunity to bail out early, without
+          // even visiting the fiber.
```

Comment:
I was trying to avoid this by also implementing changes in https://github.com/reactjs/rfcs/pull/118 because you end up with the ability to run expensive selectors multiple times if there is intermediate work that causes a bailed out fiber to end up rendering anyway.

I have no idea if "expensive selectors" is a real worry of course and so you're likely right that it wouldn't matter a whole lot perf wise.

There is one more edge case that this would trip up on though and that is if a deeper component with a selector is going to get props from an ancestor that will update on the same context change you can run the deeper selector with props that came from an earlier version of the context value.

For instance if you have a context with a key `a` and an ancestor component sees `a` and passes it to a deeper component which uses it to select from the context. then on a context value update `a` is deleted and `b` is the key. the ancestor will eventually pass `b` to the deeper child but the selector call here during propagation would try to read `a` and get an error.

Reply:
> I was trying to avoid this by also implementing changes in reactjs/rfcs#118 because you end up with the ability to run expensive selectors multiple times if there is intermediate work that causes a bailed out fiber to end up rendering anyway.

Yeah that's why I left it as a TODO 👍 Just wanted the comment there for future reference. Also why I called the dependency field "hook" since it's likely we'll need an actual reference to the current hook there, rather than a boolean.